File: src/js/fable-core/BitConverter.ts
import Long, { fromBits } from "./Long";

const littleEndian = true;

export function isLittleEndian() {
  return littleEndian;
}
export function getBytesBoolean(value: boolean) {
  const bytes = new Uint8Array(1);
  new DataView(bytes.buffer).setUint8(0, value ? 1 : 0);
  return bytes;
}
export function getBytesChar(value: string) {
  const bytes = new Uint8Array(2);
  new DataView(bytes.buffer).setUint16(0, value.charCodeAt(0), littleEndian);
  return bytes;
}
export function getBytesInt16(value: number) {
  const bytes = new Uint8Array(2);
  new DataView(bytes.buffer).setInt16(0, value, littleEndian);
  return bytes;
}
export function getBytesInt32(value: number) {
  const bytes = new Uint8Array(4);
  new DataView(bytes.buffer).setInt32(0, value, littleEndian);
  return bytes;
}
export function getBytesInt64(value: Long) {
  const bytes = new Uint8Array(8);
  new DataView(bytes.buffer).setInt32(littleEndian ? 0 : 4, value.getLowBits(), littleEndian);
  new DataView(bytes.buffer).setInt32(littleEndian ? 4 : 0, value.getHighBits(), littleEndian);
  return bytes;
}
export function getBytesUInt16(value: number) {
  const bytes = new Uint8Array(2);
  new DataView(bytes.buffer).setUint16(0, value, littleEndian);
  return bytes;
}
export function getBytesUInt32(value: number) {
  const bytes = new Uint8Array(4);
  new DataView(bytes.buffer).setUint32(0, value, littleEndian);
  return bytes;
}
export function getBytesUInt64(value: Long) {
  const bytes = new Uint8Array(8);
  new DataView(bytes.buffer).setUint32(littleEndian ? 0 : 4, value.getLowBitsUnsigned(), littleEndian);
  new DataView(bytes.buffer).setUint32(littleEndian ? 4 : 0, value.getHighBitsUnsigned(), littleEndian);
  return bytes;
}
export function getBytesSingle(value: number) {
  const bytes = new Uint8Array(4);
  new DataView(bytes.buffer).setFloat32(0, value, littleEndian);
  return bytes;
}
export function getBytesDouble(value: number) {
  const bytes = new Uint8Array(8);
  new DataView(bytes.buffer).setFloat64(0, value, littleEndian);
  return bytes;
}
export function int64BitsToDouble(value: Long) {
  const buffer = new ArrayBuffer(8);
  new DataView(buffer).setInt32(littleEndian ? 0 : 4, value.getLowBits(), littleEndian);
  new DataView(buffer).setInt32(littleEndian ? 4 : 0, value.getHighBits(), littleEndian);
  return new DataView(buffer).getFloat64(0, littleEndian);
}
export function doubleToInt64Bits(value: number) {
  const buffer = new ArrayBuffer(8);
  new DataView(buffer).setFloat64(0, value, littleEndian);
  const lowBits = new DataView(buffer).getInt32(littleEndian ? 0 : 4, littleEndian);
  const highBits = new DataView(buffer).getInt32(littleEndian ? 4 : 0, littleEndian);
  return fromBits(lowBits, highBits, false);
}
export function toBoolean(bytes: Uint8Array, offset: number): boolean {
  return new DataView(bytes.buffer).getUint8(offset) === 1 ? true : false;
}
export function toChar(bytes: Uint8Array, offset: number) {
  const code = new DataView(bytes.buffer).getUint16(offset, littleEndian);
  return String.fromCharCode(code);
}
export function toInt16(bytes: Uint8Array, offset: number) {
  return new DataView(bytes.buffer).getInt16(offset, littleEndian);
}
export function toInt32(bytes: Uint8Array, offset: number) {
  return new DataView(bytes.buffer).getInt32(offset, littleEndian);
}
export function toInt64(bytes: Uint8Array, offset: number) {
  const lowBits = new DataView(bytes.buffer).getInt32(offset + (littleEndian ? 0 : 4), littleEndian);
  const highBits = new DataView(bytes.buffer).getInt32(offset + (littleEndian ? 4 : 0), littleEndian);
  return fromBits(lowBits, highBits, false);
}
export function toUInt16(bytes: Uint8Array, offset: number) {
  return new DataView(bytes.buffer).getUint16(offset, littleEndian);
}
export function toUInt32(bytes: Uint8Array, offset: number) {
  return new DataView(bytes.buffer).getUint32(offset, littleEndian);
}
export function toUInt64(bytes: Uint8Array, offset: number) {
  const lowBits = new DataView(bytes.buffer).getUint32(offset + (littleEndian ? 0 : 4), littleEndian);
  const highBits = new DataView(bytes.buffer).getUint32(offset + (littleEndian ? 4 : 0), littleEndian);
  return fromBits(lowBits, highBits, true);
}
export function toSingle(bytes: Uint8Array, offset: number) {
  return new DataView(bytes.buffer).getFloat32(offset, littleEndian);
}
export function toDouble(bytes: Uint8Array, offset: number) {
  return new DataView(bytes.buffer).getFloat64(offset, littleEndian);
}
export function toString(bytes: Uint8Array, offset?: number, count?: number) {
  let ar = bytes;
  if (typeof offset !== "undefined" && typeof count !== "undefined") {
    ar = bytes.subarray(offset, offset + count);
  } else if (typeof offset !== "undefined") {
    ar = bytes.subarray(offset);
  }
  return Array.from(ar).map((b) => ("0" + b.toString(16)).slice(-2)).join("-");
}

File: src/fable-library-pyx/fable_library/bit_converter.py
import struct
import sys

from typing import Optional


def get_bytes_char(value: str) -> bytes:
    return bytearray(bytes(value, "UTF-8"))


def get_bytes_int16(value: int) -> bytes:
    return bytearray(value.to_bytes(length=2, byteorder=sys.byteorder))


def get_bytes_uint16(value: int) -> bytes:
    return bytearray(value.to_bytes(length=2, byteorder=sys.byteorder))


def get_bytes_int32(value: int) -> bytes:
    return bytearray(value.to_bytes(length=4, byteorder=sys.byteorder))


def get_bytes_uint32(value: int) -> bytes:
    return bytearray(value.to_bytes(length=4, byteorder=sys.byteorder))


def get_bytes_int64(value: int) -> bytes:
    return bytearray(value.to_bytes(length=8, byteorder=sys.byteorder))


def get_bytes_uint64(value: int) -> bytes:
    return bytearray(value.to_bytes(length=8, byteorder=sys.byteorder))


def get_bytes_boolean(value: bool) -> bytes:
    return bytearray(value.to_bytes(length=1, byteorder=sys.byteorder))


def get_bytes_single(value: float) -> bytes:
    return bytearray(struct.pack("f", value))


def get_bytes_double(value: float) -> bytes:
    return bytearray(struct.pack("d", value))


def int64bits_to_double(value: int) -> float:
    bytes = value.to_bytes(length=8, byteorder=sys.byteorder)
    [number] = struct.unpack("d", bytes)
    return number


def double_to_int64bits(value: float) -> int:
    bytes = bytearray(struct.pack("d", value))
    [number] = struct.unpack("q", bytes)
    return number


def to_boolean(bytes: bytearray, offset: int) -> bool:
    return bool(bool.from_bytes(bytes[offset : offset + 1], byteorder=sys.byteorder))


def to_char(bytes: bytearray, offset: int) -> str:
    return bytes[offset : offset + 1].decode("utf8")


def to_int16(bytes: bytearray, offset: int) -> int:
    return int.from_bytes(
        bytes[offset : offset + 2], byteorder=sys.byteorder, signed=True
    )


def to_uint16(bytes: bytearray, offset: int) -> int:
    return int.from_bytes(
        bytes[offset : offset + 2], byteorder=sys.byteorder, signed=False
    )


def to_int32(bytes: bytearray, offset: int) -> int:
    return int.from_bytes(
        bytes[offset : offset + 4], byteorder=sys.byteorder, signed=True
    )


def to_uint32(bytes: bytearray, offset: int) -> int:
    return int.from_bytes(
        bytes[offset : offset + 4], byteorder=sys.byteorder, signed=False
    )


def to_int64(bytes: bytes, offset: int) -> int:
    return int.from_bytes(
        bytes[offset : offset + 8], byteorder=sys.byteorder, signed=True
    )


def to_uint64(bytes: bytearray, offset: int) -> int:
    return int.from_bytes(
        bytes[offset : offset + 8], byteorder=sys.byteorder, signed=False
    )


def to_single(bytes: bytearray, offset: int) -> float:
    [number] = struct.unpack("f", bytes[offset : offset + 4])
    return number


def to_double(bytes: bytearray, offset: int) -> float:
    [number] = struct.unpack("d", bytes[offset : offset + 8])
    return number


def to_string(bytes: bytearray, offset: int = 0, count: Optional[int] = None) -> str:
    count = len(bytes) if count is None else count
    return "-".join([f"{x:02x}" for x in bytes[offset : offset + count]])


def is_little_endian() -> bool:
    return sys.byteorder == "little"


__all__ = [
    "get_bytes_char",
    "get_bytes_int16",
    "get_bytes_uint16",
    "get_bytes_int32",
    "get_bytes_uint32",
    "get_bytes_int64",
    "get_bytes_uint64",
    "get_bytes_boolean",
    "get_bytes_single",
    "get_bytes_double",
    "int64bits_to_double",
    "double_to_int64bits",
    "to_boolean",
    "to_char",
    "to_int16",
    "to_uint16",
    "to_int32",
    "to_uint64",
    "to_single",
    "to_double",
    "to_string",
    "is_little_endian",
]
